### PR TITLE
add trimming of published output based on sharedfx

### PR DESCRIPTION
**Pull request by anurse**
Any dependencies which **exactly** match the version requested in the
graph originating at the `type: platform` dependency (if any) are
trimmed from the publish output

/cc @davidfowl @piotrpMSFT @pakrym

---
This change is [Reviewable](https://reviewable.io/reviews/dotnet/cli/2097)

### CODE REVIEW

File: src/dotnet/commands/dotnet-publish/PublishCommand.cs [hunk 1]
```diff
@@ -153,7 +155,13 @@ private bool PublishProjectContext(ProjectContext context, string buildBasePath,
             var exporter = context.CreateExporter(configuration);
 
             var isPortable = string.IsNullOrEmpty(context.RuntimeIdentifier);
-            foreach (var export in exporter.GetAllExports())
+
+            // Collect all exports and organize them
+            var exports = exporter.GetAllExports()
+                .ToDictionary(e => e.Library.Identity.Name);
```
Comment (pakrym) [inline src/dotnet/commands/dotnet-publish/PublishCommand.cs]: This could be dangerous as proven by lots of `Duplicated key` errors in `DepencyModel`

**Comment by pakrym** [comment]
Having a test would make me feel better, preferably with version overriden package in it.

File: src/dotnet/commands/dotnet-publish/PublishCommand.cs (hunk 1 above)
Comment (anurse) [inline src/dotnet/commands/dotnet-publish/PublishCommand.cs]: I filtered by Package, so we shouldn't get any duplicates now.

File: src/dotnet/commands/dotnet-publish/PublishCommand.cs [hunk 2]
```diff
@@ -198,6 +207,40 @@ private bool PublishProjectContext(ProjectContext context, string buildBasePath,
...
+                return exclusionList;
+            }
+            else if (redistPackages.Count > 1)
+            {
+                throw new InvalidOperationException("Multiple packages with type: \"platform\" were specified!");
+            }
+            var redistExport = exports[redistPackages[0].Name];
+
+            exclusionList.Add(redistExport.Library.Identity.Name);
+            CollectDependencies(exports, redistExport.Library.Dependencies, exclusionList);
```
Comment (pakrym) [inline src/dotnet/commands/dotnet-publish/PublishCommand.cs]: Shouldn't `RequestedRanges` be used here?

Comment (pakrym) [inline src/dotnet/commands/dotnet-publish/PublishCommand.cs]: I thought dependencies get resolved versions of library.

**Comment by brthor** [comment]
Is this for published PortableApps? 

If an app has rids, but type platform for Microsoft.NETCore.App, would we still expect this trimming to happen?

**Comment by anurse** [comment]
> If an app has rids, but type platform for Microsoft.NETCore.App, would we still expect this trimming to happen?

Right now, yes. The behavior in that scenario is not fully defined and we can tidy up post RC2, but you would end up with a "standalone" publish, but without the runtime bits. It might fit the "shared-but-not-portable" app scenario we've spec'd out, but we'd need to go over that scenario since there may be edge cases we're missing.

File: src/dotnet/commands/dotnet-publish/PublishCommand.cs [hunk 3]
```diff
@@ -198,6 +207,40 @@ private bool PublishProjectContext(ProjectContext context, string buildBasePath,
...
+            return exclusionList;
+        }
+
+        private void CollectDependencies(Dictionary<string, LibraryExport> exports, IEnumerable<LibraryRange> dependencies, HashSet<string> exclusionList)
+        {
+            foreach (var dependency in dependencies)
+            {
+                var export = exports[dependency.Name];
+                if(export.Library.Identity.Version.Equals(dependency.VersionRange.MinVersion))
```
Comment (davidfowl) [inline src/dotnet/commands/dotnet-publish/PublishCommand.cs]: nit: spacing

File: src/dotnet/commands/dotnet-publish/PublishCommand.cs [hunk 4]
```diff
@@ -198,6 +207,40 @@ private bool PublishProjectContext(ProjectContext context, string buildBasePath,
...
+            exclusionList.Add(redistExport.Library.Identity.Name);
+            CollectDependencies(exports, redistExport.Library.Dependencies, exclusionList);
+            return exclusionList;
+        }
+
+        private void CollectDependencies(Dictionary<string, LibraryExport> exports, IEnumerable<LibraryRange> dependencies, HashSet<string> exclusionList)
+        {
+            foreach (var dependency in dependencies)
+            {
+                var export = exports[dependency.Name];
```
Comment (davidfowl) [inline src/dotnet/commands/dotnet-publish/PublishCommand.cs]: Can this ever throw? What about unresovled dependencies?

**Pull request merged**